x-pack/platform/plugins/private/upgrade_assistant/server/lib/data_streams/data_stream_reindex_service.ts

```diff
@@ -0,0 +1,287 @@
+            )}`
+          );
+        }
+
+        if (taskResponse.complete) {
+          // Check that no failures occurred
+          if (taskResponse.errors.length) {
+            // Include the entire task result in the error message. This should be guaranteed
+            // to be JSON-serializable since it just came back from Elasticsearch.
+            throw error.reindexTaskFailed(
+              `Reindexing failed with ${taskResponse.errors.length} errors:\n${JSON.stringify(
+                taskResponse,
+                null,
+                2
+              )}`
+            );
+          }
+
+          // Update the status
+          return {
+            reindexTaskPercComplete: 1,
+            status: DataStreamReindexStatus.completed,
+            progressDetails: {
+              successCount: taskResponse.successes,
+              pendingCount: taskResponse.pending,
+              inProgressCount: (taskResponse.in_progress ?? []).length,
+              errorsCount: (taskResponse.errors ?? []).length,
+            },
+          };
+        } else {
+          // Updated the percent complete
+          const perc = taskResponse.successes / taskResponse.total_indices_in_data_stream;
+
+          return {
+            status: DataStreamReindexStatus.inProgress,
+            reindexTaskPercComplete: perc,
+            progressDetails: {
+              successCount: taskResponse.successes,
+              pendingCount: taskResponse.pending,
+              inProgressCount: (taskResponse.in_progress ?? []).length,
+              errorsCount: (taskResponse.errors ?? []).length,
+            },
+          };
+        }
+      } catch (err) {
+        if (
+          err.name === 'ResponseError' &&
+          (err.message as string).includes('resource_not_found_exception')
+        ) {
+          // cancelled, never started, or successful task but finished from than 24 hours ago
+          // Since this API should be called as a follow up from _migrate API, we can assume that the task is not started
+          return {
+            status: DataStreamReindexStatus.notStarted,
+          };
+        }
+
+        return {
+          status: DataStreamReindexStatus.failed,
+          errorMessage: err.toString(),
+        };
+      }
+    },
+    async cancelReindexing(dataStreamName: string) {
+      const resp = await esClient.transport.request<{ acknowledged: boolean }>({
+        method: 'POST',
+        path: `/_migration/reindex/${dataStreamName}/_cancel`,
+      });
+
+      if (!resp.acknowledged) {
+        throw error.reindexCannotBeCancelled(`Could not cancel reindex.`);
+      }
+
+      return {
+        status: DataStreamReindexStatus.cancelled,
+      };
+    },
+    async getDataStreamMetadata(dataStreamName: string): Promise<DataStreamMetadata> {
+      try {
+        const { body: statsBody } = (await esClient.transport.request(
+          {
+            method: 'GET',
+            path: `/${dataStreamName}/_stats`,
```

Comment:
The value being returned in `lastBackingIndexCreationDate` is the newest backing index in the data stream. We should return the newest backing index with incompatible data.

For this metadata we need to use [deprecations info API](https://www.elastic.co/guide/en/elasticsearch/reference/current/migration-api-deprecation.html). Then use the `indicesRequiringUpgrade: string[]`, only checking `settings.index.creation_date` for those.